### PR TITLE
#439: Added default support for environment-driven configuration.

### PR DESCRIPTION
**Pull request by schrodingersket**
This pull request addresses #439 and provides environment-driven support which respects existing file mounts and utilizes existing environment configuration from the official Sentry Docker image. This is achieved in two ways:

1) The primary configuration folder is moved from `/etc/sentry` -> `/etc/sentry/onpremise`. This allows the built-in config files to be used (which are environment-driven). 

2) Out-of-the-box configuration is provided by including `docker-compose.env.yml`, which contains defaults to run with exactly the same configuration as provided in the example config files.

### CODE REVIEW

File: docker-compose.override.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+version: '3.4'
```
Comment (BYK) [inline docker-compose.override.yml:1]: I'm guessing you didn't intend to include this file in your commit? :)

File: docker-compose.yml [hunk 2]
```diff
@@ -20,11 +20,13 @@ x-sentry-defaults: &sentry_defaults
     - symbolicator
     - kafka
   environment:
-    SENTRY_CONF: '/etc/sentry'
+    SENTRY_CONF: '/etc/sentry/onpremise'
     SNUBA: 'http://snuba-api:1218'
   volumes:
     - 'sentry-data:/data'
-    - './sentry:/etc/sentry'
+    - './sentry/config.yml:/etc/sentry/onpremise/config.yml'
```
Comment (BYK) [inline docker-compose.yml:27]: Shall we switch to https://docs.docker.com/compose/compose-file/#configs as we are changing this part?

File: docker-compose.yml [hunk 3]
```diff
@@ -164,8 +166,6 @@ services:
     command: '"0 0 * * * gosu sentry sentry cleanup --days $SENTRY_EVENT_RETENTION_DAYS"'
   nginx:
     << : *restart_policy
-    ports:
```
Comment (BYK) [inline docker-compose.yml:167]: Revert this change please :)

File: docker-compose.env.yml [hunk 4]
```diff
@@ -0,0 +1,131 @@
+version: '3.4'
+
+# All environment variables referenced from
```
Comment (BYK) [inline docker-compose.env.yml:3]: I have 2 issues with this file:

1. It should not replicate _everything_ we have in the original `docker-compose.yml` file as maintaining parity between them would be a nightmare.
2. The list of environment variables getting out of sync with the original config file. Any idea how to simplify this?

Comment (schrodingersket) [reply docker-compose.env.yml:3]: > 1. It should not replicate everything we have in the original docker-compose.yml file as maintaining parity between them would be a nightmare.

I think I'd mentioned in #439 only adding support for the connection information (host/port/db name) for the infra services... would that be acceptable? I think at the very least, avoiding configuration for anything that can be configured in-app is probably wise, yeah?

> 2. The list of environment variables getting out of sync with the original config file. Any idea how to simplify this?

I think in addressing point 1, that's probably mitigated since only core infrastructural config that's unlikely to change is exposed. What do you think?

File: docker-compose.override.yml (hunk 1 above)
Comment (schrodingersket) [reply docker-compose.override.yml:1]: Initially, I'd meant to; IIRC, when I'd created this branch initially, the port was exposed on the `web` service, which implied that the port exposure was a setting not meant for production (e.g., in production, that port would be closed off in favor of running behind a proxy, and the port mapping would be in developer overrides in `docker-compose.override.yml`). 

The main reason I'd added this in the first place is that an override file can't remove an existing port mapping, and wanted the public port exposure to be optional for the fairly common use-case of running behind a proxy without exposing a redundant port.

What are your thoughts on that? If you'd like to keep that mapping in the main `docker-compose.yml` file, would you be willing to make the publicly-exposed port configurable?

File: docker-compose.yml (hunk 2 above)
Comment (schrodingersket) [reply docker-compose.yml:27]: Sure! I assume we'd point to the config files generated after running `install.sh`, yeah?

**Comment by schrodingersket** [comment]
What's the best way to add tests for this configuration? I admit that I'm not terribly familiar with Travis CI, but the test file here seems pretty straightforward. Would running the same test script as for the config file install be acceptable?

Thanks as always for your time on this!!

File: docker-compose.yml (hunk 2 above)
Comment (schrodingersket) [reply docker-compose.yml:27]: @BYK - looks like the `configs` feature is only supported in Docker Swarm: https://github.com/docker/compose/issues/5110

Shall we leave this as a folder mount?

File: docker-compose.env.yml (hunk 4 above)
Comment (BYK) [reply docker-compose.env.yml:3]: >I think I'd mentioned in #439 only adding support for the connection information (host/port/db name) for the infra services... would that be acceptable? I think at the very least, avoiding configuration for anything that can be configured in-app is probably wise, yeah?

Yup, that's easier. I'm wondering if we can somehow automate this. Worth thinking hard on it as that would make maintenance incredibly easier. Maybe establishing a pattern like `SENTRY_CONF_*` or something. Still though, I'm finding this harder to reason about the more implementation details I see:

 - If we need to list these in a file, then why not just list them in an `.env` file, with their values?
 - And then if we are listing them in a _file_, why can we not make that file the config file we already have?

I guess with tests + minimal connection options, this can be okay.

**Comment by BYK** [comment]
>What's the best way to add tests for this configuration? I admit that I'm not terribly familiar with Travis CI, but the test file here seems pretty straightforward. Would running the same test script as for the config file install be acceptable?

I think at this point we'd need multiple jobs as defined here: https://docs.travis-ci.com/user/build-matrix/#explicitly-included-jobs-with-only-one-element-in-the-build-matrix

In the second job definition, we'd forcefully use the ENV-driven config and set some values and see if they actually work. That said, this probably means we should not rely on docker-compose to set other services up and then set them up manually. At this point, I again start to question whether this approach makes sense or not.

>Thanks as always for your time on this!!

Thank _you_ for driving this and pushing forward. I know I may sound negative/pessimistic at times but I really want to understand the use case and it's longer-term fit into the project while making this work for you and people similar to your use case. Your contributions are very valuable and much appreciated :)

**Comment by BYK** [comment]
@schrodingersket it's been quite a while and I was going through some old issues and PRs that got stale. Is there any way to salvage the work you've already put here or should we close this PR?

Either way, thanks a lot for all your time and effort on this PR and the related issue!

**Comment by schrodingersket** [comment]
@BYK - should be good to close this guy out. I think the complexity of the Docker Compose set up warrants simply installing this guy as a standalone deployment and integrating that way, rather than integrating it into other projects via composition of Docker Compose files. (: Thanks for all the awesome feedback though - definitely helped me put my own configuration and setup through its paces!

**Comment by BYK** [comment]
@schrodingersket thanks! Happy to work on another PR that we can merge in the future :)
